awsx/ecr/auth.ts

```diff
@@ -0,0 +1,93 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as aws from "@pulumi/aws";
+import * as pulumi from "@pulumi/pulumi";
+
+/**
+ * Arguments for fetching ECR registry credentials.
+ */
+export interface CredentialArgs {
+  /**
+   * The URL of the ECR registry to get credentials for.
+   * Can be provided with or without the https:// protocol prefix.
+   */
+  registryUrl: string;
+
+  /**
+   * Optional registry ID (AWS account ID) to get credentials for.
+   * If not provided, will be parsed from the registry URL.
+   */
+  registryId?: string;
+}
+
+/**
+ * Docker registry credentials for authenticating with an ECR registry.
+ */
+export interface DockerCredentials {
+  /**
+   * The address of the ECR registry.
+   */
+  address: string;
+
+  /**
+   * The username to authenticate with. For ECR this is typically "AWS".
+   */
+  username: string;
+
+  /**
+   * The password to authenticate with. For ECR this is a temporary token.
+   */
+  password: string;
+}
+
+/**
+ * Fetches Docker registry credentials for authenticating with an ECR registry.
+ *
+ * @param args Arguments for fetching ECR registry credentials
+ * @param opts InvokeOutputOptions to use for the credential lookup
+ * @returns Docker registry credentials including address, username and password
+ */
+export function getDockerCredentials(
+  args: CredentialArgs,
+  opts: pulumi.InvokeOutputOptions,
+): pulumi.Output<DockerCredentials> {
+  let registryId: string;
+  if (args.registryId) {
+    registryId = args.registryId;
+  } else {
+    // add protocol to help parse the url
+    const registryUrl = args.registryUrl?.startsWith("https://")
+      ? args.registryUrl
+      : "https://" + args.registryUrl;
+    const parsedUrl = new URL(registryUrl);
+    // the registry id is the AWS account id. It's the first part of the hostname
+    registryId = parsedUrl.hostname.split(".")[0];
```

Comment:
Should this get a bit more error handling just in case, or we are certain it succeeds?

Reply:
I've taken this from the existing implementation of `Image`, but you have a good point!
What could fail in theory is parsing the URL (i.e. `new URL(...)`), doesn't hurt to add some error handling for that to return better errors.

Reply:
Added error handling + tests for it